dbt/adapters/sql/connections.py

```diff
@@ -199,3 +211,46 @@ def commit(self):
         connection.transaction_open = False
 
         return connection
+
+    def _retryable_cursor_execute(self,
+                                  execute_fn: Callable,
+                                  sql: str,
+                                  bindings: Optional[Any] = None,
+                                  retryable_exceptions: Iterable[Type[Exception]] = [],
+                                  retry_limit: int = 1,
+                                  retry_timeout: Union[Callable[[int], SleepTime], SleepTime] = 1,
+                                  _attempts: int = 0,
+                                  ) -> None:
+        timeout = retry_timeout(_attempts) if callable(retry_timeout) else retry_timeout
+        if timeout < 0:
+            raise DbtRuntimeError(
+                "retry_timeout cannot be negative or return a negative time."
+            )
+
+        try:
+            execute_fn(sql, bindings)
+
+        except tuple(retryable_exceptions) as e:
+            retry_limit -= 1
+            if retry_limit <= 0:
+                raise e
+            fire_event(
+                AdapterEventDebug(
+                    message=f"Got a retryable error {type(e)} when attempting to execute a query.\n"
+                            f"{retry_limit} attempts remaining. Retrying in {timeout} seconds.\n"
+                            f"Error:\n{e}"
+                )
+            )
+
+            time.sleep(timeout)
+            return self._retryable_cursor_execute(
+                execute_fn=execute_fn,
+                sql=sql,
+                retry_limit=retry_limit - 1,
+                retry_timeout=retry_timeout,
+                retryable_exceptions=retryable_exceptions,
+                _attempts=_attempts + 1,
```

Comment:
I see why we would pass in `_attempts` now, but this could be a separate concept, similar to the way BQ manages retries. I still think it's an early abstraction and we could update this with a more intelligent method in a future update.